drivers/hmis/app/models/hmis/hud/custom_client_name.rb

```diff
@@ -52,13 +52,14 @@ def full_name
   end
 
   def update_client_name
-    client.update(
-      first_name: first,
-      last_name: last,
-      middle_name: middle,
-      name_suffix: suffix,
-      name_data_quality: name_data_quality || 99,
-    )
+    # Removed calling update from within an after_save callback is dangerous(?) and caused warehouse processing job not to trigger for Client
```

Comment:
I think the problem with calling `update` within an `after_save` hook, is that it will cause a `save` event, so can loop out of control.  
What about calling it in `before_save` and just setting the attributes?
Alternatively, you could dirty the record with a new column) and then have a background process that looks for dirty records every few minutes and fixes them up.

Reply:
I like the idea of a separate process that reconciles them, but it doesn't help in cases where the propagation really should happen immediately (like during a mutation).

I considered the `before_save` option, but it doesn't feel right since then the caller would then be expected to call `name.client.save!` after, which is unlikely. The other direction-- before_save on the Client-- doesn't seem quite right either, since you also need to call it in a specific way.

There are only 2 places where this propagation is needed, so I don't think it's _too_ much to just ask the caller to handle it. (Which is what I propose in this pr with `assign_primary_name_fields`).
1) When a Client is created through an application form submission. This always goes through the `ClientProcessor` which handles the name propagation.
2) The Referral creation API endpoint, which creates a CustomClientName for a new Client.
~~3) Client Merge~~ it isn't actually needed here, since we only mark a name as primary if it already matches the hud Client's name fields.